### PR TITLE
JS: Small performance improvements.

### PR DESCRIPTION
**Pull request by erik-krogh**
I tried to figure out the performance regression in https://github.com/github/codeql/pull/3391, and that lead me to find some completely unrelated performance improvements. Which is why I'm opening a separate PR for these performance improvements. 

8716790 and f8de691 helps reduce the impact of https://github.com/github/codeql/pull/3391, and 6d05b40 just improves performance in general. 

An [evaluation on nightly](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1589190090216) shows about 5% performance improvement. 
(A [redo of the all the regressions](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1589195382488), and [yet another redo](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1589200162307) of just the worst offender). 

I have confirmed that each commit on its own improve performance, although the main impact seems to come from 6d05b40. 
Here is an evaluation where master is compared first with just 6d05b40 and with all 3 commits together: [link](https://docs.google.com/spreadsheets/d/1AlN78l0ukbBhS6yX0cRS0P2wfmrkBqzMv_A1YLKEHNA/edit#gid=0) (in a sheet because formatting [got weird](https://git.semmle.com/erik/dist-compare-reports/tree/erik-laptop_1589154083950) in the report). 

6d05b40 can cause some single queries to be slower, as it forces the `cached ReachableBasicBlock::strictlyDominates` predicate to be evaluated, and the query might not have needed that predicate without my change.   
But as `strictlyDominates` is a cached predicate, that performance penalty completely disappears when running many queries.

TODO: 
- [x] Big-apps evaluation

### CODE REVIEW

**Comment by asgerf** [comment]
Neat! I'd say these changes looked sensible even without the performance boost 👍

File: javascript/ql/src/semmle/javascript/InclusionTests.qll [hunk 1]
```diff
@@ -71,8 +71,7 @@ module InclusionTest {
       count(this.getACallee()) = 1 and
       count(callee.getAReturnedExpr()) = 1 and
       not this.isImprecise() and
-      inner.getContainerNode().getALocalSource().getEnclosingExpr() = callee.getAParameter() and
-      inner.getContainedNode().getALocalSource().getEnclosingExpr() = callee.getAParameter()
+      inner.getContainerNode().getALocalSource() = DataFlow::parameterNode(callee.getAParameter())
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/InclusionTests.qll:74]: I do not understand this change. 
The previous version used both `getContainedNode` and `getContainerNode`.

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/InclusionTests.qll:74]: You are right, that was a mistake from my end, I didn't notice the `Contained`/`Container`. 

Everything ends up working anyway, as `getContainedNode` is called all the places the `InclusionTest` is used. And thus the `getContainedNode` constraint still enforced, just not in the CharPred.

I fixed it anyway (and refactored two other `getEnclosingExpr()`).

**Comment by esbena** [comment]
Great. We are just waiting for the evaluation then.

**Comment by erik-krogh** [comment]
A [big-apps evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-esben.northeurope.cloudapp.azure.com_1589282363674) looks very good. 

All evaluations were without the latest change to `IndirectInclusionTest`, so I did [a smoke-test on just the changes to `IndirectInclusionTest`](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1589288754154), and that looks fine. 

I think this is ready to land.

**Pull request merged**